src/online2bin/online2-tcp-nnet3-decode-faster.cc

```diff
@@ -0,0 +1,457 @@
+
+    KALDI_LOG << "Loading FST...";
+
+    fst::Fst<fst::StdArc> *decode_fst = ReadFstKaldiGeneric(fst_rxfilename);
+
+    fst::SymbolTable *word_syms = NULL;
+    if (!word_syms_rxfilename.empty())
+      if (!(word_syms = fst::SymbolTable::ReadText(word_syms_rxfilename)))
+        KALDI_ERR << "Could not read symbol table from file "
+                  << word_syms_rxfilename;
+
+    KALDI_LOG << "Loaded evertyhing. Waiting for data...";
+
+    signal(SIGPIPE, SIG_IGN); //ignore SIGPIPE to avoid crashing when socket forecfully disconnected
+
+    TcpServer server(read_timeout);
+
+    server.Listen(port_num);
+
+    while (true) {
+
+      server.Accept();
+
+      int32 samp_count = 0;//this is used for output refresh rate
+      size_t chunk_len = static_cast<size_t>(chunk_length_secs * samp_freq);
+
+      int32 check_update = static_cast<int32>(samp_freq * output_freq);
+      int32 next_check = check_update;
+
+      int32 frame_offset = 0;
+
+      bool eos = false;
+
+      OnlineIvectorExtractorAdaptationState adaptation_state(
+          feature_info.ivector_extractor_info);
+
+      OnlineNnet2FeaturePipeline feature_pipeline(feature_info);
+      SingleUtteranceNnet3Decoder decoder(decoder_opts, trans_model,
+                                          decodable_info,
+                                          *decode_fst, &feature_pipeline);
+
+      while (!eos) {
+
+        decoder.InitDecoding(frame_offset);
+
+        OnlineSilenceWeighting silence_weighting(
+            trans_model,
+            feature_info.silence_weighting_config,
+            decodable_opts.frame_subsampling_factor);
+
+        std::vector<std::pair<int32, BaseFloat>> delta_weights;
+        
+        while (true) {
+
+          eos = !server.ReadChunk(chunk_len);
+
+          if (!eos) {
+            Vector<BaseFloat> wave_part = server.GetChunk();
+            feature_pipeline.AcceptWaveform(samp_freq, wave_part);
+            samp_count += chunk_len;
+
+            if (silence_weighting.Active() &&
+                feature_pipeline.IvectorFeature() != NULL) {
+              silence_weighting.ComputeCurrentTraceback(decoder.Decoder());
+              silence_weighting.GetDeltaWeights(feature_pipeline.NumFramesReady(),
+                                                &delta_weights);
+              feature_pipeline.UpdateFrameWeights(delta_weights,
+                                                  frame_offset * decodable_opts.frame_subsampling_factor);
+            }
+
+            decoder.AdvanceDecoding();
+
+            if (samp_count > next_check) {
+              if (decoder.NumFramesDecoded() > 0) {
+                Lattice lat;
+                decoder.GetBestPath(false, &lat);
+
+                std::string msg = LatticeToString(lat, word_syms);
+
+                server.Write(msg + '\r');
+              }
+
+              next_check += check_update;
+            }
+
+          } else {
```

Comment:
I think it would be better/simpler to change this else clause to `if (eos)` and move it above since it ends with a break statement. So the new logic would be

```C++
while (true) {
  eos = !server.ReadChunk(chunk_len);

  if (eos) {
    // handle end of stream
    break;
  }

  Vector<BaseFloat> wave_part = server.GetChunk();
  // handle chunk

  if (decoder.EndpointDetected(endpoint_opts)) {
    // handle endpoint
  }
}
```

Reply:
Yea, I'll have to figure out how to test this. The models that I have don't rely on silence weighing.

I'll fix the code as suggested.